File: collatz.hs
-- Tested with GHCi 8.0.2
import System.Environment

collatz :: Int -> [Int]
collatz n 
  | (n == 1) = [1]
  | n `mod` 2 == 0 = n:collatz (n `div` 2)
  | otherwise = n:collatz (3 * n + 1)


main :: IO ()
main =
  do
    args <- map read <$> getArgs
    putStrLn $ unlines $ map show $ collatz $ args !! 0

File: collatz.pl
% Tested with SWI-Prolog (threaded, 64 bits, version 9.2.9)
collatz(1,1) :- writeln(1), !.
collatz(N,R) :- M is N mod 2,  M = 0, !, writeln(N), D is N div 2, collatz(D,R).
collatz(N,R) :- !, M is 3 * N + 1, writeln(N), collatz(M,R).		  
